### PR TITLE
SAK-41260 Accessibility date picker and calendar widget once open with keyboard

### PR DESCRIPTION
**Pull request by gwickram**
https://sakaiproject.atlassian.net/browse/SAK-41260

### CODE REVIEW

File: library/src/webapp/js/lang-datepicker/lang-datepicker.js [hunk 1]
```diff
@@ -1592,22 +1627,17 @@ $.datepicker._gotoToday = function (id) {
...
-		// on select, runs our custom method for setting dates
-		cfg.onSelect = function(dtObj, dpInst) {
-			setHiddenFields($(this).datepicker("getDate"), options, dpInst);
-      if (options.onDateTimeSelected) {
-        options.onDateTimeSelected(moment($(this).datepicker("getDate")).valueOf());
-      }
-		};
+                }
+
+                if (options.ashidden !== undefined) {
```
Comment (adrianfish) [inline library/src/webapp/js/lang-datepicker/lang-datepicker.js:1637]:
```suggestion
                if (options.ashidden) {
```

File: library/src/webapp/js/lang-datepicker/lang-datepicker.js [hunk 2]
```diff
@@ -1436,6 +1436,14 @@ if(c&&c._defaults.timeOnly&&b.input.val()!==b.lastVal)try{$.datepicker._updateDa
...
 
+$.datepicker._updateDatepicker_original = $.datepicker._updateDatepicker;
+$.datepicker._updateDatepicker = function(inst) {
+	$.datepicker._updateDatepicker_original(inst);
+	var afterShow = this._get(inst, 'afterShow');
```
Comment (adrianfish) [inline library/src/webapp/js/lang-datepicker/lang-datepicker.js:1442]: Use const here. When we start linting these files var will fail the lint.

File: library/src/webapp/js/lang-datepicker/lang-datepicker.js [hunk 3]
```diff
@@ -1570,8 +1579,34 @@ $.datepicker._gotoToday = function (id) {
...
+                                }, 100);
+                };
+
+
 		// add blur event to allow edit input field
-		cfg.beforeShow = function(input, inst) {
+		// Moved to init to handle updates without date-picker popup
+                //cfg.beforeShow = function(input, inst) {
+                //	setBlur(input);
+                //};
```
Comment (adrianfish) [inline library/src/webapp/js/lang-datepicker/lang-datepicker.js:1594]:
```suggestion
```

File: library/src/webapp/js/lang-datepicker/lang-datepicker.js [hunk 4]
```diff
@@ -1570,8 +1579,34 @@ $.datepicker._gotoToday = function (id) {
...
+                cfg.onSelect = function(dtObj, dpInst) {
+                        setHiddenFields($(this).datepicker("getDate"), options, dpInst);
+                        if (options.onDateTimeSelected) {
+                                options.onDateTimeSelected(moment($(this).datepicker("getDate")).valueOf());
+                        }
+                };
+
+                // When the picker allows empty dates, it should detect when the date is removed from the input, and update the hidden value.
+                cfg.onClose = function(dtObj, dpInst) {
+                        if (dtObj == '' && options.allowEmptyDate){
```
Comment (adrianfish) [inline library/src/webapp/js/lang-datepicker/lang-datepicker.js:1605]: I'd use the triple = here.

File: library/src/webapp/js/lang-datepicker/lang-datepicker.js [hunk 5]
```diff
@@ -1592,22 +1627,17 @@ $.datepicker._gotoToday = function (id) {
 						$(input).val(stringDay);
 					}
 				}
+				// Set hidden field to null if allowed
+                                if (options.allowEmptyDate && $(input).val() == "") {
```
Comment (adrianfish) [inline library/src/webapp/js/lang-datepicker/lang-datepicker.js:1631]: I'd clean up the wrapping of input here. We can just wrap it once in the setBlur function.

File: library/src/webapp/js/lang-datepicker/lang-datepicker.js [hunk 6]
```diff
@@ -1721,6 +1751,13 @@ $.datepicker._gotoToday = function (id) {
 							  newValue = moment(d).format();
 							  break;
 						}
+                                        } else if (d == "") {
```
Comment (adrianfish) [inline library/src/webapp/js/lang-datepicker/lang-datepicker.js:1754]: triple === is better

File: library/src/webapp/js/lang-datepicker/lang-datepicker.js [hunk 7]
```diff
@@ -1721,6 +1751,13 @@ $.datepicker._gotoToday = function (id) {
...
+                                             newValue = null;
+                                        }
+                                        // If oldvalue is empty, we need to update the hidden field twice, to unlock the submit
+                                        // button in datemanager
+                                        if (oldValue == '' && newValue != '') {
```
Comment (adrianfish) [inline library/src/webapp/js/lang-datepicker/lang-datepicker.js:1759]:
```suggestion
                                        if (oldValue === '' && newValue) {
```

File: library/src/webapp/js/lang-datepicker/lang-datepicker.js [hunk 8]
```diff
@@ -1697,7 +1727,7 @@ $.datepicker._gotoToday = function (id) {
 				jQuery.each(o.ashidden, function(i, h) {
 					var oldValue = jQuery('#' + h).val();
 					var newValue = '';
-					if(d != null){
+					if(d != null && d != ""){
```
Comment (adrianfish) [inline library/src/webapp/js/lang-datepicker/lang-datepicker.js:1730]:
```suggestion
					if (d) {
```

**Review by adrianfish — CHANGES_REQUESTED** [review]
A few changes. I think what you're trying to achieve here is to set a timeout that focuses the date picker after it has displayed, is that right?

**Comment by adrianfish** [comment]
Please request  a re review when you're ready, and I'll get on it. Thanks for the work :)

**Review by ottenhoff — CHANGES_REQUESTED** [review]

File: library/src/webapp/js/lang-datepicker/lang-datepicker.js (hunk 2 above)
Comment (gwickram) [reply library/src/webapp/js/lang-datepicker/lang-datepicker.js]: It allows keyboard accessibility, not forcing the calendar to popup when typing in the text field, and set proper focus on the first input control when it pops up with the button.

**Comment by gwickram** [comment]
The improvements were added. Thanks.

File: library/src/webapp/js/lang-datepicker/lang-datepicker.js (hunk 2 above)
Comment (ottenhoff) [reply library/src/webapp/js/lang-datepicker/lang-datepicker.js]: Yes, that is the aim of the PR. But I think our question on the core call was about these specific lines. What is line 1439 doing where the original is copied over?

Comment (ottenhoff) [reply library/src/webapp/js/lang-datepicker/lang-datepicker.js]: Is there a simpler or more clear way to do this?

**Pull request merged**